// angular-auth-oidc-client/esm2015/lib/login/login.service.js
import { Injectable } from '@angular/core';
import { ConfigurationProvider } from '../config/config.provider';
import { ParLoginService } from './par/par-login.service';
import { PopUpLoginService } from './popup/popup-login.service';
import { StandardLoginService } from './standard/standard-login.service';
export class LoginService {
    constructor(configurationProvider, parLoginService, popUpLoginService, standardLoginService) {
        this.configurationProvider = configurationProvider;
        this.parLoginService = parLoginService;
        this.popUpLoginService = popUpLoginService;
        this.standardLoginService = standardLoginService;
    }
    login(authOptions) {
        const { usePushedAuthorisationRequests } = this.configurationProvider.getOpenIDConfiguration();
        if (usePushedAuthorisationRequests) {
            return this.parLoginService.loginPar(authOptions);
        }
        else {
            return this.standardLoginService.loginStandard(authOptions);
        }
    }
    loginWithPopUp(authOptions, popupOptions) {
        const { usePushedAuthorisationRequests } = this.configurationProvider.getOpenIDConfiguration();
        if (usePushedAuthorisationRequests) {
            return this.parLoginService.loginWithPopUpPar(authOptions, popupOptions);
        }
        else {
            return this.popUpLoginService.loginWithPopUpStandard(authOptions, popupOptions);
        }
    }
}
LoginService.decorators = [
    { type: Injectable }
];
LoginService.ctorParameters = () => [
    { type: ConfigurationProvider },
    { type: ParLoginService },
    { type: PopUpLoginService },
    { type: StandardLoginService }
];
