// angular-auth-oidc-client/esm2015/lib/auto-login/auto-login.guard.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { of } from 'rxjs';
import { concatMap, map } from 'rxjs/operators';
import { AuthStateService } from '../authState/auth-state.service';
import { AutoLoginService } from '../auto-login/auto-login-service';
import { CheckAuthService } from '../check-auth.service';
import { LoginService } from '../login/login.service';
import * as i0 from "@angular/core";
import * as i1 from "./auto-login-service";
import * as i2 from "../authState/auth-state.service";
import * as i3 from "../check-auth.service";
import * as i4 from "../login/login.service";
import * as i5 from "@angular/router";
export class AutoLoginGuard {
    constructor(autoLoginService, authStateService, checkAuthService, loginService, router) {
        this.autoLoginService = autoLoginService;
        this.authStateService = authStateService;
        this.checkAuthService = checkAuthService;
        this.loginService = loginService;
        this.router = router;
    }
    canLoad(route, segments) {
        return this.checkAuth(route.path);
    }
    canActivate(route, state) {
        return this.checkAuth(state.url);
    }
    checkAuth(url) {
        return this.authStateService.authorized$.pipe(concatMap((isAuthenticatedAlready) => (isAuthenticatedAlready ? of(isAuthenticatedAlready) : this.checkAuthService.checkAuth())), map((isAuthorized) => {
            const storedRoute = this.autoLoginService.getStoredRedirectRoute();
            if (isAuthorized) {
                if (storedRoute) {
                    this.autoLoginService.deleteStoredRedirectRoute();
                    this.router.navigate([storedRoute]);
                }
                return true;
            }
            this.autoLoginService.saveStoredRedirectRoute(url);
            this.loginService.login();
            return false;
        }));
    }
}
AutoLoginGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function AutoLoginGuard_Factory() { return new AutoLoginGuard(i0.ɵɵinject(i1.AutoLoginService), i0.ɵɵinject(i2.AuthStateService), i0.ɵɵinject(i3.CheckAuthService), i0.ɵɵinject(i4.LoginService), i0.ɵɵinject(i5.Router)); }, token: AutoLoginGuard, providedIn: "root" });
AutoLoginGuard.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
AutoLoginGuard.ctorParameters = () => [
    { type: AutoLoginService },
    { type: AuthStateService },
    { type: CheckAuthService },
    { type: LoginService },
    { type: Router }
];
