// angular-auth-oidc-client/esm2015/lib/login/popup/popup.service.js
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class PopUpService {
    constructor() {
        this.STORAGE_IDENTIFIER = 'popupauth';
        this.receivedUrlInternal$ = new Subject();
    }
    get receivedUrl$() {
        return this.receivedUrlInternal$.asObservable();
    }
    isCurrentlyInPopup() {
        const popup = sessionStorage.getItem(this.STORAGE_IDENTIFIER);
        return !!window.opener && window.opener !== window && !!popup;
    }
    openPopUp(url, popupOptions) {
        const optionsToPass = this.getOptions(popupOptions);
        this.popUp = window.open(url, '_blank', optionsToPass);
        this.popUp.sessionStorage.setItem(this.STORAGE_IDENTIFIER, 'true');
        const listener = (event) => {
            if (!(event === null || event === void 0 ? void 0 : event.data) || typeof event.data !== 'string') {
                return;
            }
            this.receivedUrlInternal$.next(event.data);
            this.cleanUp(listener);
        };
        window.addEventListener('message', listener, false);
    }
    sendMessageToMainWindow(url) {
        if (window.opener) {
            this.sendMessage(url, window.location.href);
        }
    }
    cleanUp(listener) {
        window.removeEventListener('message', listener, false);
        if (this.popUp) {
            this.popUp.sessionStorage.removeItem(this.STORAGE_IDENTIFIER);
            this.popUp.close();
            this.popUp = null;
        }
    }
    sendMessage(url, href) {
        window.opener.postMessage(url, href);
    }
    getOptions(popupOptions) {
        const popupDefaultOptions = { width: 500, height: 500, left: 50, top: 50 };
        const options = Object.assign(Object.assign({}, popupDefaultOptions), (popupOptions || {}));
        return Object.entries(options)
            .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`)
            .join(',');
    }
}
PopUpService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PopUpService_Factory() { return new PopUpService(); }, token: PopUpService, providedIn: "root" });
PopUpService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
