// angular-auth-oidc-client/esm2015/lib/utils/flowHelper/flow-helper.service.js
import { Injectable } from '@angular/core';
import { ConfigurationProvider } from '../../config/config.provider';
export class FlowHelper {
    constructor(configurationProvider) {
        this.configurationProvider = configurationProvider;
    }
    isCurrentFlowCodeFlow() {
        return this.currentFlowIs('code');
    }
    isCurrentFlowAnyImplicitFlow() {
        return this.isCurrentFlowImplicitFlowWithAccessToken() || this.isCurrentFlowImplicitFlowWithoutAccessToken();
    }
    isCurrentFlowCodeFlowWithRefreshTokens() {
        const { useRefreshToken } = this.configurationProvider.getOpenIDConfiguration();
        if (this.isCurrentFlowCodeFlow() && useRefreshToken) {
            return true;
        }
        return false;
    }
    isCurrentFlowImplicitFlowWithAccessToken() {
        return this.currentFlowIs('id_token token');
    }
    isCurrentFlowImplicitFlowWithoutAccessToken() {
        return this.currentFlowIs('id_token');
    }
    currentFlowIs(flowTypes) {
        const { responseType } = this.configurationProvider.getOpenIDConfiguration();
        if (Array.isArray(flowTypes)) {
            return flowTypes.some((x) => responseType === x);
        }
        return responseType === flowTypes;
    }
}
FlowHelper.decorators = [
    { type: Injectable }
];
FlowHelper.ctorParameters = () => [
    { type: ConfigurationProvider }
];
