// angular-auth-oidc-client/esm2015/lib/callback/code-flow-callback.service.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { throwError } from 'rxjs';
import { catchError, tap } from 'rxjs/operators';
import { ConfigurationProvider } from '../config/config.provider';
import { FlowsDataService } from '../flows/flows-data.service';
import { FlowsService } from '../flows/flows.service';
import { IntervallService } from './intervall.service';
import * as i0 from "@angular/core";
import * as i1 from "../flows/flows.service";
import * as i2 from "../flows/flows-data.service";
import * as i3 from "./intervall.service";
import * as i4 from "../config/config.provider";
import * as i5 from "@angular/router";
export class CodeFlowCallbackService {
    constructor(flowsService, flowsDataService, intervallService, configurationProvider, router) {
        this.flowsService = flowsService;
        this.flowsDataService = flowsDataService;
        this.intervallService = intervallService;
        this.configurationProvider = configurationProvider;
        this.router = router;
    }
    authorizedCallbackWithCode(urlToCheck) {
        const isRenewProcess = this.flowsDataService.isSilentRenewRunning();
        const { triggerAuthorizationResultEvent, postLoginRoute, unauthorizedRoute } = this.configurationProvider.getOpenIDConfiguration();
        return this.flowsService.processCodeFlowCallback(urlToCheck).pipe(tap((callbackContext) => {
            if (!triggerAuthorizationResultEvent && !callbackContext.isRenewProcess) {
                this.router.navigate([postLoginRoute]);
            }
        }), catchError((error) => {
            this.flowsDataService.resetSilentRenewRunning();
            this.intervallService.stopPeriodicallTokenCheck();
            if (!triggerAuthorizationResultEvent && !isRenewProcess) {
                this.router.navigate([unauthorizedRoute]);
            }
            return throwError(error);
        }));
    }
}
CodeFlowCallbackService.ɵprov = i0.ɵɵdefineInjectable({ factory: function CodeFlowCallbackService_Factory() { return new CodeFlowCallbackService(i0.ɵɵinject(i1.FlowsService), i0.ɵɵinject(i2.FlowsDataService), i0.ɵɵinject(i3.IntervallService), i0.ɵɵinject(i4.ConfigurationProvider), i0.ɵɵinject(i5.Router)); }, token: CodeFlowCallbackService, providedIn: "root" });
CodeFlowCallbackService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
CodeFlowCallbackService.ctorParameters = () => [
    { type: FlowsService },
    { type: FlowsDataService },
    { type: IntervallService },
    { type: ConfigurationProvider },
    { type: Router }
];
