// angular-auth-oidc-client/esm2015/lib/callback/implicit-flow-callback.service.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { throwError } from 'rxjs';
import { catchError, tap } from 'rxjs/operators';
import { ConfigurationProvider } from '../config/config.provider';
import { FlowsDataService } from '../flows/flows-data.service';
import { FlowsService } from '../flows/flows.service';
import { IntervallService } from './intervall.service';
import * as i0 from "@angular/core";
import * as i1 from "../flows/flows.service";
import * as i2 from "../config/config.provider";
import * as i3 from "@angular/router";
import * as i4 from "../flows/flows-data.service";
import * as i5 from "./intervall.service";
export class ImplicitFlowCallbackService {
    constructor(flowsService, configurationProvider, router, flowsDataService, intervalService) {
        this.flowsService = flowsService;
        this.configurationProvider = configurationProvider;
        this.router = router;
        this.flowsDataService = flowsDataService;
        this.intervalService = intervalService;
    }
    authorizedImplicitFlowCallback(hash) {
        const isRenewProcess = this.flowsDataService.isSilentRenewRunning();
        const { triggerAuthorizationResultEvent, postLoginRoute, unauthorizedRoute } = this.configurationProvider.getOpenIDConfiguration();
        return this.flowsService.processImplicitFlowCallback(hash).pipe(tap((callbackContext) => {
            if (!triggerAuthorizationResultEvent && !callbackContext.isRenewProcess) {
                this.router.navigate([postLoginRoute]);
            }
        }), catchError((error) => {
            this.flowsDataService.resetSilentRenewRunning();
            this.intervalService.stopPeriodicallTokenCheck();
            if (!triggerAuthorizationResultEvent && !isRenewProcess) {
                this.router.navigate([unauthorizedRoute]);
            }
            return throwError(error);
        }));
    }
}
ImplicitFlowCallbackService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ImplicitFlowCallbackService_Factory() { return new ImplicitFlowCallbackService(i0.ɵɵinject(i1.FlowsService), i0.ɵɵinject(i2.ConfigurationProvider), i0.ɵɵinject(i3.Router), i0.ɵɵinject(i4.FlowsDataService), i0.ɵɵinject(i5.IntervallService)); }, token: ImplicitFlowCallbackService, providedIn: "root" });
ImplicitFlowCallbackService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
ImplicitFlowCallbackService.ctorParameters = () => [
    { type: FlowsService },
    { type: ConfigurationProvider },
    { type: Router },
    { type: FlowsDataService },
    { type: IntervallService }
];
