// angular-auth-oidc-client/esm2015/lib/callback/intervall.service.js
import { Injectable, NgZone } from '@angular/core';
import { Observable } from 'rxjs';
import * as i0 from "@angular/core";
export class IntervallService {
    constructor(zone) {
        this.zone = zone;
        this.runTokenValidationRunning = null;
    }
    stopPeriodicallTokenCheck() {
        if (this.runTokenValidationRunning) {
            this.runTokenValidationRunning.unsubscribe();
            this.runTokenValidationRunning = null;
        }
    }
    startPeriodicTokenCheck(repeatAfterSeconds) {
        const millisecondsDelayBetweenTokenCheck = repeatAfterSeconds * 1000;
        return new Observable((subscriber) => {
            let intervalId;
            this.zone.runOutsideAngular(() => {
                intervalId = setInterval(() => subscriber.next(), millisecondsDelayBetweenTokenCheck);
            });
            return () => {
                clearInterval(intervalId);
            };
        });
    }
}
IntervallService.ɵprov = i0.ɵɵdefineInjectable({ factory: function IntervallService_Factory() { return new IntervallService(i0.ɵɵinject(i0.NgZone)); }, token: IntervallService, providedIn: "root" });
IntervallService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
IntervallService.ctorParameters = () => [
    { type: NgZone }
];
