// angular-auth-oidc-client/esm2015/lib/callback/refresh-session-refresh-token.service.js
import { Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { FlowsService } from '../flows/flows.service';
import { ResetAuthDataService } from '../flows/reset-auth-data.service';
import { LoggerService } from '../logging/logger.service';
import { IntervallService } from './intervall.service';
import * as i0 from "@angular/core";
import * as i1 from "../logging/logger.service";
import * as i2 from "../flows/reset-auth-data.service";
import * as i3 from "../flows/flows.service";
import * as i4 from "./intervall.service";
export class RefreshSessionRefreshTokenService {
    constructor(loggerService, resetAuthDataService, flowsService, intervalService) {
        this.loggerService = loggerService;
        this.resetAuthDataService = resetAuthDataService;
        this.flowsService = flowsService;
        this.intervalService = intervalService;
    }
    refreshSessionWithRefreshTokens(customParams) {
        this.loggerService.logDebug('BEGIN refresh session Authorize');
        return this.flowsService.processRefreshToken(customParams).pipe(catchError((error) => {
            this.intervalService.stopPeriodicallTokenCheck();
            this.resetAuthDataService.resetAuthorizationData();
            return throwError(error);
        }));
    }
}
RefreshSessionRefreshTokenService.ɵprov = i0.ɵɵdefineInjectable({ factory: function RefreshSessionRefreshTokenService_Factory() { return new RefreshSessionRefreshTokenService(i0.ɵɵinject(i1.LoggerService), i0.ɵɵinject(i2.ResetAuthDataService), i0.ɵɵinject(i3.FlowsService), i0.ɵɵinject(i4.IntervallService)); }, token: RefreshSessionRefreshTokenService, providedIn: "root" });
RefreshSessionRefreshTokenService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
RefreshSessionRefreshTokenService.ctorParameters = () => [
    { type: LoggerService },
    { type: ResetAuthDataService },
    { type: FlowsService },
    { type: IntervallService }
];
