// angular-auth-oidc-client/esm2015/lib/iframe/refresh-session-iframe.service.js
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable, RendererFactory2 } from '@angular/core';
import { Observable } from 'rxjs';
import { LoggerService } from '../logging/logger.service';
import { UrlService } from '../utils/url/url.service';
import { SilentRenewService } from './silent-renew.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../logging/logger.service";
import * as i3 from "../utils/url/url.service";
import * as i4 from "./silent-renew.service";
export class RefreshSessionIframeService {
    constructor(doc, loggerService, urlService, silentRenewService, rendererFactory) {
        this.doc = doc;
        this.loggerService = loggerService;
        this.urlService = urlService;
        this.silentRenewService = silentRenewService;
        this.renderer = rendererFactory.createRenderer(null, null);
    }
    refreshSessionWithIframe(customParams) {
        this.loggerService.logDebug('BEGIN refresh session Authorize Iframe renew');
        const url = this.urlService.getRefreshSessionSilentRenewUrl(customParams);
        return this.sendAuthorizeRequestUsingSilentRenew(url);
    }
    sendAuthorizeRequestUsingSilentRenew(url) {
        const sessionIframe = this.silentRenewService.getOrCreateIframe();
        this.initSilentRenewRequest();
        this.loggerService.logDebug('sendAuthorizeRequestUsingSilentRenew for URL:' + url);
        return new Observable((observer) => {
            const onLoadHandler = () => {
                sessionIframe.removeEventListener('load', onLoadHandler);
                this.loggerService.logDebug('removed event listener from IFrame');
                observer.next(true);
                observer.complete();
            };
            sessionIframe.addEventListener('load', onLoadHandler);
            sessionIframe.contentWindow.location.replace(url);
        });
    }
    initSilentRenewRequest() {
        const instanceId = Math.random();
        const initDestroyHandler = this.renderer.listen('window', 'oidc-silent-renew-init', (e) => {
            if (e.detail !== instanceId) {
                initDestroyHandler();
                renewDestroyHandler();
            }
        });
        const renewDestroyHandler = this.renderer.listen('window', 'oidc-silent-renew-message', (e) => this.silentRenewService.silentRenewEventHandler(e));
        this.doc.defaultView.dispatchEvent(new CustomEvent('oidc-silent-renew-init', {
            detail: instanceId,
        }));
    }
}
RefreshSessionIframeService.ɵprov = i0.ɵɵdefineInjectable({ factory: function RefreshSessionIframeService_Factory() { return new RefreshSessionIframeService(i0.ɵɵinject(i1.DOCUMENT), i0.ɵɵinject(i2.LoggerService), i0.ɵɵinject(i3.UrlService), i0.ɵɵinject(i4.SilentRenewService), i0.ɵɵinject(i0.RendererFactory2)); }, token: RefreshSessionIframeService, providedIn: "root" });
RefreshSessionIframeService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
RefreshSessionIframeService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: LoggerService },
    { type: UrlService },
    { type: SilentRenewService },
    { type: RendererFactory2 }
];
