// angular-auth-oidc-client/esm2015/lib/config/auth-well-known.service.js
import { Injectable } from '@angular/core';
import { of, throwError } from 'rxjs';
import { catchError, tap } from 'rxjs/operators';
import { EventTypes } from '../public-events/event-types';
import { PublicEventsService } from '../public-events/public-events.service';
import { StoragePersistanceService } from '../storage/storage-persistance.service';
import { AuthWellKnownDataService } from './auth-well-known-data.service';
export class AuthWellKnownService {
    constructor(publicEventsService, dataService, storagePersistanceService) {
        this.publicEventsService = publicEventsService;
        this.dataService = dataService;
        this.storagePersistanceService = storagePersistanceService;
    }
    getAuthWellKnownEndPoints(authWellknownEndpointUrl) {
        const alreadySavedWellKnownEndpoints = this.storagePersistanceService.read('authWellKnownEndPoints');
        if (!!alreadySavedWellKnownEndpoints) {
            return of(alreadySavedWellKnownEndpoints);
        }
        return this.getWellKnownEndPointsFromUrl(authWellknownEndpointUrl).pipe(tap((mappedWellKnownEndpoints) => this.storeWellKnownEndpoints(mappedWellKnownEndpoints)), catchError((error) => {
            this.publicEventsService.fireEvent(EventTypes.ConfigLoadingFailed, null);
            return throwError(error);
        }));
    }
    storeWellKnownEndpoints(mappedWellKnownEndpoints) {
        this.storagePersistanceService.write('authWellKnownEndPoints', mappedWellKnownEndpoints);
    }
    getWellKnownEndPointsFromUrl(authWellknownEndpoint) {
        return this.dataService.getWellKnownEndPointsFromUrl(authWellknownEndpoint);
    }
}
AuthWellKnownService.decorators = [
    { type: Injectable }
];
AuthWellKnownService.ctorParameters = () => [
    { type: PublicEventsService },
    { type: AuthWellKnownDataService },
    { type: StoragePersistanceService }
];
