// angular-auth-oidc-client/esm2015/lib/login/par/par.service.js
import { HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { catchError, map, retry } from 'rxjs/operators';
import { DataService } from '../../api/data.service';
import { LoggerService } from '../../logging/logger.service';
import { StoragePersistanceService } from '../../storage/storage-persistance.service';
import { UrlService } from '../../utils/url/url.service';
export class ParService {
    constructor(loggerService, urlService, dataService, storagePersistanceService) {
        this.loggerService = loggerService;
        this.urlService = urlService;
        this.dataService = dataService;
        this.storagePersistanceService = storagePersistanceService;
    }
    postParRequest(customParams) {
        let headers = new HttpHeaders();
        headers = headers.set('Content-Type', 'application/x-www-form-urlencoded');
        const authWellKnown = this.storagePersistanceService.read('authWellKnownEndPoints');
        if (!authWellKnown) {
            return throwError('Could not read PAR endpoint because authWellKnownEndPoints are not given');
        }
        const parEndpoint = authWellKnown.parEndpoint;
        if (!parEndpoint) {
            return throwError('Could not read PAR endpoint from authWellKnownEndpoints');
        }
        const data = this.urlService.createBodyForParCodeFlowRequest(customParams);
        return this.dataService.post(parEndpoint, data, headers).pipe(retry(2), map((response) => {
            this.loggerService.logDebug('par response: ', response);
            return {
                expiresIn: response.expires_in,
                requestUri: response.request_uri,
            };
        }), catchError((error) => {
            const errorMessage = `There was an error on ParService postParRequest`;
            this.loggerService.logError(errorMessage, error);
            return throwError(errorMessage);
        }));
    }
}
ParService.decorators = [
    { type: Injectable }
];
ParService.ctorParameters = () => [
    { type: LoggerService },
    { type: UrlService },
    { type: DataService },
    { type: StoragePersistanceService }
];
